### ISSUE TITLE
Can Not Create Snapshot and the Whole Backup Hanging / May Need to Add a Timeout Protection Mechanism When Creating Snapshot

### ISSUE THREAD

**Issue by AnhaoROMA** [VictoriaMetrics/VictoriaMetrics#3551]
# General Idea

I was confronted with such a situation that the vmbackup-process hung because of failing to create the snapshot. And it did cause some trouble especially when I set vmbackup as a cronjob to do backup regularly. 

This phenomenon has happened several times, so I want to know the reason why creating snapshot fails. If creating snapshot can be affected by some episodic factors, it might be better to add a timeout protection mechanism, from my point of view.

# Describe the Bug

On the 27th of December, I found the vmbackup-process had existed for over 3 days, which is not a normal case.
<img width="935" alt="hang_for_3_days" src="https://user-images.githubusercontent.com/77609772/209685166-817d4dcb-8f64-4a66-be0f-93202ec235c2.png">

After checking the log file, I thought it was caused by the failure to create snapshot. And this process exists / hangs since that.
<img width="937" alt="cannot_create_snapshot" src="https://user-images.githubusercontent.com/77609772/209685792-a82f989a-bbfd-4bf6-b477-d58d32ce91df.png">

Then I tried to create the snapshot manually, but unfortunately it failed!
<img width="659" alt="cannot_wget_createURL" src="https://user-images.githubusercontent.com/77609772/209686062-0e7005db-85e0-47a0-902e-47af0e03f8f8.png">

However, in comparison, it succeeded when I tried to delete_all snapshot(s) manually.
<img width="890" alt="can_wget_deleteURL" src="https://user-images.githubusercontent.com/77609772/209686204-025af615-1d96-41aa-bfa6-2c8df9332570.png">

By reading the code in [snapshot.go](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/snapshot/snapshot.go) , I found it lacks the timeout protection mechanism when creating snapshot.
<img width="474" alt="code_to_create_snapshot" src="https://user-images.githubusercontent.com/77609772/209686453-18b010b7-51e9-4cb2-ba17-f513af5df3a6.png">

In my opinion, when vmbackup is set to run regularly, and the new vmbackup-process will not start unless the old vmbackup-process has completed, it will cause some trouble.

# Expected Behavior
1. Give me some help to know the reason why creating snapshot fail sometimes.
2. If creating snapshot can be affected by some occasional / episodic / incidental factors, is it reasonable to add a timeout protection mechanism?

**Comment by dmitryk-dk** [comment]
Hi @AnhaoROMA ! We shuold improve behavior of the vmbackup

**Comment by AnhaoROMA** [comment]
Thanks to @dmitryk-dk very much!

**Comment by zekker6** [comment]
@AnhaoROMA Aside from enhancing error handling at vmbackup side it would be great for you to check if your VM installation is healthy. 
Usually creating snapshot is very fast(under 1m) as it is based on building hardlinks references for existing partitions without an actual data copying. In your case it seems to be taking much more time which might be a sign of something going not too well with an installation.
It would be great if you could setup [monitoring](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#monitoring) for your installation and use our [alerts set(`alerts*.yaml` files)](https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/deployment/docker) to monitor it.

**Comment by AnhaoROMA** [comment]
@zekker6 Will try. Thank you!

**Comment by zekker6** [comment]
Closing this as resolved, feel free to reopen if you will have any other related questions.

**Comment by k-stz** [comment]
We're seeing this exact issue as well. Stack: K8s, deploying CR `vmsingle` with victoriametrics operator. 

Why is this closed? Timeout protection doesn't solve the issue that the snapshot creation is hanging on the victormetrics instance itself. 

**When snapshot creation hangs, further attempts hang as well**:
Though the api `snapshot/delete_all` works, a call to `snapshot/create` hangs indefinetly. On the client side (when using `curl vm-url:port/snapshot/create`) I get a `504 Gateway Time-out` while the victoriametrics-vmsingle logs show:
```logs
2023-01-27T15:21:09.553Z        info    VictoriaMetrics/lib/storage/storage.go:318      creating Storage snapshot for "/path-to-vm-storage"
```

To recover from the hanging snapshots I need to restart the database. Any other ideas, we're doing hourly backups, would be great to have a call that could reset the API and make snapshot creation work again.

## What we tried:
- Testing for 2 months now, moved database to its own disk but the sporadic snapshot hanging keeps showing up. 
- Update: to vmbackup and victorametrics image `1.79.5`

## Expected behaviour:
In my opinion the database should recover from this on it's own. Even if it might be a VM/baremetall issue, why not recover from this and then try the 1-millisecond-snapshot-creation again. The database still ingests metrics properly, so it can't be anything permanently affecting the disk I/O.

## Questions
Any ideas of how to recover from this automatically, is this a software issue? Why can't the db recover on its own? Only after a restarting does the db creating snapshots works again

**Comment by AnhaoROMA** [comment]
@k-stz 
Yes, by restarting VictoriaMetrics can recover the situation.

**Comment by AnhaoROMA** [comment]
As [#3858](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3858) has been solved, this issue can be closed.

**Comment by zekker6** [comment]
@AnhaoROMA I'm reopening this as we're going to implement 2 more fixes to make this more consistent:
- add a timeout protection to force cancelling snapshots if before next backup will start
- deleting incomplete backups when any error occurs

Also it would be great to have metrics to track failed snapshots.

Once this will be done we'll close this issue.

**Comment by AnhaoROMA** [comment]
@zekker6 Thanks a lot!

**Comment by valyala** [comment]
The `-snapshotCreateTimeout` command-line flag is available starting from [v1.88.1](https://docs.victoriametrics.com/CHANGELOG.html).
VictoriaMetrics also deletes incomplete backups starting from `v1.88.1`.

Closing the issue as done.

**Comment by AnhaoROMA** [comment]
@valyala Thanks a lot!

**Comment by valyala** [comment]
It has been appeared that the snapshot process could take long time because the loop inside the `flushInmemoryParts()` function could be potentially infinite under high data ingestion rate. This has been fixed in the commit 6460475e3bb488086046cb5fdce3915206d731ef . This commit will be included in the next release. Re-opening the issue until the next release with the proper fix.

**Comment by valyala** [comment]
The problem mentioned in the previous comment has been fixed in [VictoriaMetrics v1.87.4 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.4). Closing the issue as fixed.

**Comment by valyala** [comment]
FYI, the issue with slow [snapshot creating](https://docs.victoriametrics.com/#how-to-work-with-snapshots) has been also fixed starting from [VictoriaMetrics v1.90.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
